### DISCUSSION TITLE
How to determine the number of variables in runtime?

### DISCUSSION THREAD

**Discussion by dzhliu** [autodiff/autodiff#211] [Q&A]
To calculate the partial derivative, we need to construct a function:

var f(var x, var y, var z)
{
    return 1 + x + y + z + x*y + y*z + x*z + x*y*z + exp(x/y + y/z);
}
Then, we call
auto [ux, uy, uz] = derivatives(u, wrt(x, y, z)); 

But when the equation is a black-box model and the number of variables can only be determined in runtime, we can't define the function before compiling. In this situation, how to call wrt() and how to construct the function?

**Comment by allanleal** [comment]
Hi. If your function is a black-box and cannot be changed to use `autodiff` number types, then it is not possible to compute its derivatives with automatic differentiation.

**Reply by dzhliu** [reply]
Thanks for your prompt reply.
But what I mean is, the number of variables can only be determined dynamically. For example, the function could be:
y=ax+by
But then, it could also be:
y=ax+by+cz
In this situation, we can't pass variables to "wrt" function manually, since the number of variables differs with time goes by. So if it is possible to dynamically do this in runtime?

**Reply by allanleal** [reply]
In this case, you may want to define a scalar function with a vector as argument and use `gradient` to compute the derivatives. See if this example helps: https://autodiff.github.io/tutorials/#gradient-of-a-scalar-function

**Reply by dzhliu** [reply]
Thanks for your help. Now Autodiff works smoothly in my project.

The last question about Autodiff is, if it is possible to calculate gradient in Matrix style?

I mean, traditionally we implement this in scalar style. For example, given a constant value **_a_** and a variable **_x_**, we can calculate the derivative of **_y_** toward **_x_**, namely, 
y=ax 
dydx = a

But if it is possible to implement this process in matrix style? Which means, given **_n_** constants and one single variable **_x_**, we have the equation:
y=[a1,a2,...,an]x
If we calculate the partial derivative of **_y_** toward **_x_**, we can get:
dydx = [a1,a2,...,an]

Thanks!

**Reply by allanleal** [reply]
I'm not sure I understood your need. Do you need the Jacobian of a vector-valued function? If so, this is currently supported only for `real` and `dual` types. Check the tutorials for Jacobian computation: https://autodiff.github.io/tutorials/

**Reply by dzhliu** [reply]
I already tried Jacobian. But there exist another problem: reference to non-static member function must be called. I can give you a example here:

class func{ 
   ArrayXreal f(ArrayXreal x, ArrayXreal c)
   {
          return x*c;
   }
};

int main()
{
       func fu();
       initial c as ArrayXreal and fill it with data
       initial x as ArrayXreal and fill it with data
       jacobian(fu.f, wrt(x,c),at(x,c));
}

Now you will get the error: **reference to non-static member function must be called**

That's to say, the first parameter (function pointer) of Jacobian must be pointed to a static function. I am really confused about this idea. I am wondering if the function f must be static?

**Reply by allanleal** [reply]
Please provide complete working examples. Please also use markdown to properly render your codes.
